web/src/lib.rs

```diff
@@ -1,60 +1,291 @@
+
         Ok(())
     }
 
-    fn clear(&mut self) {
-        // TODO: this should probably be async too.
-        self.entries.clear()
+    async fn clear(&mut self) -> Result<(), ()> {
+        let _result_js = self.transaction(IdbTransactionMode::Readwrite, move |object_store| {
+            object_store.clear().unwrap()
+        }).await.unwrap();
+
+        Ok(())
     }
 }
 
+impl IndexedDBStore {
+    async fn new(name: &str) -> Self {
+        // TODO: handle errors properly, don't just unwrap/expect.
+        let db = Self::open_db(name).await.unwrap();
+        db.close();
+        Self { name: name.to_string() }
+    }
+
+    async fn open_db(name: &str) -> Result<IdbDatabase, ()> {
+        let window = web_sys::window().expect("No global `window` exists.");
+        let indexed_db = window.indexed_db().unwrap().expect("Should have `indexedDB` on `window`, running in an outdated browser?");
+        let open_db_request = indexed_db.open_with_u32(name, 1).unwrap();
+
+        let on_upgrade_needed_closure: Closure<dyn FnMut(_)> = Closure::once(move |event: Event| {
+            let open_db_request_js = JsValue::from(event.current_target());
+            let open_db_request = IdbOpenDbRequest::from(open_db_request_js);
+            let db_js = open_db_request.result().unwrap();
+            let db = IdbDatabase::from(db_js);
+            db.create_object_store("entries").unwrap();
+        });
+
+        let (on_error_sender, on_error_receiver) = oneshot::channel::<Result<(), ()>>();
+        let (on_success_sender, on_success_receiver) = oneshot::channel::<Result<(), ()>>();
+
+        let on_error_closure: Closure<dyn FnMut(_)> = Closure::once(move |_event: Event| {
+            // TODO: extract proper error object.
+            on_error_sender.send(Err(())).unwrap();
+        });
+
+        let on_success_closure: Closure<dyn FnMut(_)> = Closure::once(move |_event: Event| {
+            on_success_sender.send(Ok(())).unwrap();
+        });
+
+        open_db_request.set_onupgradeneeded(on_upgrade_needed_closure.as_ref().dyn_ref());
+        open_db_request.set_onerror(on_error_closure.as_ref().dyn_ref());
+        open_db_request.set_onsuccess(on_success_closure.as_ref().dyn_ref());
+
+        let on_error_receiver_fused = on_error_receiver.fuse();
+        let on_success_receiver_fused = on_success_receiver.fuse();
+        pin_mut!(on_error_receiver_fused, on_success_receiver_fused);
+
+        let result: Result<(), ()> = select! {
+            // TODO: pass proper error.
+            _ = on_error_receiver_fused => Err(()),
+            _ = on_success_receiver_fused => Ok(()),
+        };
+
+        result.expect("An error occured while opening indexed DB.");
+
+        open_db_request.set_onupgradeneeded(None);
+        open_db_request.set_onerror(None);
+        open_db_request.set_onsuccess(None);
+
+        drop(on_upgrade_needed_closure);
+        drop(on_error_closure);
+        drop(on_success_closure);
```

Comment:
I believe these `drop` statements can be removed, as the closures are dropped at the end of the function

Reply:
Yep, wasn't sure. Fixed.